astroid/bases.py

```diff
@@ -436,7 +436,14 @@ def _infer_builtin_new(
         node_context = context.extra_context.get(caller.args[0])
         infer = caller.args[0].infer(context=node_context)
 
-        yield from (Instance(x) if x is not Uninferable else x for x in infer)  # type: ignore[misc,arg-type]
+        for inferred in infer:
+            if isinstance(inferred, type(Uninferable)):  # isinstance helps mypy
+                yield inferred
+            if isinstance(inferred, Instance):
+                raise InferenceError
+            if isinstance(inferred, (nodes.ClassDef, nodes.Lambda, Proxy)):
+                yield Instance(inferred)
+            raise InferenceError
```

Comment:
If we got for the `isinstance(inferred, nodes.ClassDef)` variant we should probably switch those `if` statements around and remove the `Instance` check: we can just raise if it is not `ClassDef`.

Reply:
Thanks, I should have noticed I was getting the type from the wrong parent!